File: Jiho/week6/day28/개인정보 수집 유효기간.js
function solution(today, terms, privacies) {
  const answer = [];
  const dates = [];

  function toDateValue(dateStr) {
    const [y, m, d] = dateStr.split(".").map(Number);
    return y * 12 * 28 + m * 28 + d;
  }

  // 1. 유효기간 정보 객체로 가공(O(1)로 접근하기 위해)
  const termObj = {};
  for (const term of terms) {
    const [key, value] = term.split(" ");
    termObj[key] = Number(value);
  }

  // 2. 파기 날짜 계산
  for (let i = 0; i < privacies.length; i++) {
    let [date, type] = privacies[i].split(" ");
    let [year, month, day] = date.split(".").map((v) => Number(v));

    const tempMonth = month + termObj[type];
    year += Math.floor((tempMonth - 1) / 12);
    month = ((tempMonth - 1) % 12) + 1;

    date = [year, month, day].join(".");

    // 3. 파기날짜 지났으면 바로바로 answer에 추가
    if (toDateValue(date) <= toDateValue(today)) answer.push(i + 1);
  }

  return answer;
}

File: Jin/week6/day28/개인정보수집유효기간.java
package Jin.week6.day28;

import java.util.*;

public class 개인정보수집유효기간 {
    public int[] solution(String today, String[] terms, String[] privacies) {
        int todayDays = convertToDays(today);

        HashMap<String, Integer> termMap = new HashMap<>();

        for (String t : terms) {
            String[] termArr = t.split(" ");
            termMap.put(termArr[0], Integer.parseInt(termArr[1]));
        }

        ArrayList<Integer> result = new ArrayList<>();
        for (int i = 0; i < privacies.length; i++) {
            String[] privArr = privacies[i].split(" ");
            String date = privArr[0];
            String term = privArr[1];

            int expireDays = convertToDays(date) + (termMap.get(term) * 28);

            if (todayDays >= expireDays) {
                result.add(i + 1);
            }
        }

        return result.stream().mapToInt(Integer::intValue).toArray();
    }

    private int convertToDays(String date) {
        String[] parts = date.split("\\.");
        int year = Integer.parseInt(parts[0]);
        int month = Integer.parseInt(parts[1]);
        int day = Integer.parseInt(parts[2]);

        return (year * 12 * 28) + ((month - 1) * 28) + day;
    }
}
